File: frontend/src/lib/api/landProfile.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient, ensureCsrfCookie } from './client'

export type YesNoUnknown = 'yes' | 'no' | 'in_process' | 'unknown'
export type RiskLevel = 'none' | 'low' | 'medium' | 'high' | 'unknown'

export interface LandProfile {
  kitta_number: string | null
  lalpurja_available: YesNoUnknown
  lalpurja_document_url: string | null
  road_access: boolean
  road_width_meters: number | null
  road_type: 'blacktop' | 'gravel' | 'dirt' | 'none'
  water_access: 'municipal' | 'well' | 'none' | 'unknown'
  electricity_access: boolean
  drainage_access: 'yes' | 'no' | 'unknown'
  land_classification: 'residential' | 'agricultural' | 'commercial' | 'guthi' | 'other'
  flood_risk: RiskLevel
  landslide_risk: RiskLevel
  nearby_development_notes: string | null
  document_verification_status: 'unverified' | 'partial' | 'verified'
  verified_at: string | null
  completeness_percent: number
}

export type LandProfileInput = Partial<Omit<LandProfile, 'lalpurja_document_url' | 'document_verification_status' | 'verified_at' | 'completeness_percent'>>

export function useLandProfile(propertyId: number | undefined) {
  return useQuery({
    queryKey: ['land-profile', propertyId],
    queryFn: async () => {
      const { data } = await apiClient.get<{ data: LandProfile }>(`/properties/${propertyId}/land-profile`)
      return data.data
    },
    enabled: !!propertyId,
  })
}

export function useSaveLandProfile() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async ({ propertyId, input }: { propertyId: number; input: LandProfileInput }) => {
      await ensureCsrfCookie()
      const { data } = await apiClient.put<{ data: LandProfile }>(`/properties/${propertyId}/land-profile`, input)
      return data.data
    },
    onSuccess: (_data, variables) => {
      // The listing detail page embeds the land profile inside its own query
      // (['listings', 'detail', slug]) rather than reading useLandProfile, so
      // that cache needs invalidating too or the checklist/admin panel go stale.
      queryClient.invalidateQueries({ queryKey: ['land-profile', variables.propertyId] })
      queryClient.invalidateQueries({ queryKey: ['listings'] })
      queryClient.invalidateQueries({ queryKey: ['owner', 'properties'] })
    },
  })
}

export function useUploadLandDocument() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async ({ propertyId, file }: { propertyId: number; file: File }) => {
      const form = new FormData()
      form.append('file', file)
      await ensureCsrfCookie()
      const { data } = await apiClient.post<{ data: LandProfile }>(`/properties/${propertyId}/land-profile/document`, form, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return data.data
    },
    onSuccess: (_data, variables) => {
      // The listing detail page embeds the land profile inside its own query
      // (['listings', 'detail', slug]) rather than reading useLandProfile, so
      // that cache needs invalidating too or the checklist/admin panel go stale.
      queryClient.invalidateQueries({ queryKey: ['land-profile', variables.propertyId] })
      queryClient.invalidateQueries({ queryKey: ['listings'] })
      queryClient.invalidateQueries({ queryKey: ['owner', 'properties'] })
    },
  })
}

export function useVerifyLandProfile() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async ({ propertyId, status }: { propertyId: number; status: 'unverified' | 'partial' | 'verified' }) => {
      await ensureCsrfCookie()
      const { data } = await apiClient.patch<{ data: LandProfile }>(`/admin/properties/${propertyId}/land-profile/verify`, {
        document_verification_status: status,
      })
      return data.data
    },
    onSuccess: (_data, variables) => {
      // The listing detail page embeds the land profile inside its own query
      // (['listings', 'detail', slug]) rather than reading useLandProfile, so
      // that cache needs invalidating too or the checklist/admin panel go stale.
      queryClient.invalidateQueries({ queryKey: ['land-profile', variables.propertyId] })
      queryClient.invalidateQueries({ queryKey: ['listings'] })
      queryClient.invalidateQueries({ queryKey: ['owner', 'properties'] })
    },
  })
}

File: mobile/lib/features/listings/data/models/land_profile.dart
import '../../../../core/network/json_parsing.dart';
import '../../../../core/network/media_url.dart';

/// Mirrors `LandProfileResource` — only present when `property_type == 'land'`.
/// See backend/app/Domain/Calculators/Services/AreaUnitConverter.php for the
/// unit conversions used to compute `Property.area.display` alongside this.
class LandProfile {
  LandProfile({
    required this.kittaNumber,
    required this.lalpurjaAvailable,
    required this.lalpurjaDocumentUrl,
    required this.roadAccess,
    required this.roadWidthMeters,
    required this.roadType,
    required this.waterAccess,
    required this.electricityAccess,
    required this.drainageAccess,
    required this.landClassification,
    required this.floodRisk,
    required this.landslideRisk,
    required this.nearbyDevelopmentNotes,
    required this.documentVerificationStatus,
    required this.verifiedAt,
    required this.completenessPercent,
  });

  factory LandProfile.fromJson(Map<String, dynamic> json) {
    return LandProfile(
      kittaNumber: json['kitta_number'] as String?,
      lalpurjaAvailable: json['lalpurja_available'] as String,
      lalpurjaDocumentUrl: json['lalpurja_document_url'] != null
          ? resolveMediaUrl(json['lalpurja_document_url'] as String)
          : null,
      roadAccess: json['road_access'] as bool? ?? false,
      roadWidthMeters: asDouble(json['road_width_meters']),
      roadType: json['road_type'] as String,
      waterAccess: json['water_access'] as String,
      electricityAccess: json['electricity_access'] as bool? ?? false,
      drainageAccess: json['drainage_access'] as String,
      landClassification: json['land_classification'] as String,
      floodRisk: json['flood_risk'] as String,
      landslideRisk: json['landslide_risk'] as String,
      nearbyDevelopmentNotes: json['nearby_development_notes'] as String?,
      documentVerificationStatus: json['document_verification_status'] as String,
      verifiedAt: json['verified_at'] as String?,
      completenessPercent: asDoubleOr(json['completeness_percent'], 0),
    );
  }

  final String? kittaNumber;
  final String lalpurjaAvailable; // yes | no | in_process | unknown
  final String? lalpurjaDocumentUrl;
  final bool roadAccess;
  final double? roadWidthMeters;
  final String roadType; // blacktop | gravel | dirt | none
  final String waterAccess; // municipal | well | none | unknown
  final bool electricityAccess;
  final String drainageAccess; // yes | no | unknown
  final String landClassification;
  final String floodRisk; // none | low | medium | high | unknown
  final String landslideRisk;
  final String? nearbyDevelopmentNotes;
  final String documentVerificationStatus; // unverified | partial | verified
  final String? verifiedAt;
  final double completenessPercent;
}
